### PR TITLE
Some updates to the API for archived repos

### PR DESCRIPTION
**Pull request by JakobDev**
While playing around with the API for archived repos I found some nits that this PR fixes.
1. If a repo is archived, the API no longer returns a generic 404. Instead it returns a 423 (locked) with a meaningful error.
2. Some endpoints were missing the check if a repo is archived. I fixed that.
3. I also think it is a good idea to check in the underlying functions, if a repo is archived. If you now call e.g. `ApplyDiffPatch` for an archived repo, it now returns a `ErrRepoIsArchived` error. Those functions should never be called if a repo is archived, so you should never get this error in theory, but it's better to make sure it doesn't happen.

This PR makes no claim to completeness. I may have missed some function.

### CODE REVIEW

File: models/repo/repo.go [hunk 1]
```diff
@@ -654,6 +664,14 @@ func (repo *Repository) GetTrustModel() TrustModelType {
 	return trustModel
 }
 
+// GetIsArchivedError returns ErrRepoIsArchived is the repo is archived
```
Comment (delvh) [inline models/repo/repo.go]:
```suggestion
// GetIsArchivedError returns ErrRepoIsArchived if the repo is archived
```

File: models/repo/repo.go [hunk 2]
```diff
@@ -654,6 +664,14 @@ func (repo *Repository) GetTrustModel() TrustModelType {
...
+func (repo *Repository) GetIsArchivedError() error {
```
Comment (delvh) [inline models/repo/repo.go]: `ErrorIfArchived` or `MustNotBeArchived`?

File: models/repo/repo.go [hunk 3]
```diff
@@ -654,6 +664,14 @@ func (repo *Repository) GetTrustModel() TrustModelType {
...
+	if repo.IsArchived {
+		return ErrRepoIsArchived{ID: repo.ID, Name: repo.Name, Owner: repo.OwnerName}
```
Comment (delvh) [inline models/repo/repo.go]: I recommend against doing that, instead pass the repo directly, and then use `repo.LogString()` inside the error message.

File: modules/context/api.go [hunk 4]
```diff
@@ -101,6 +101,12 @@ type APIRedirect struct{}
 // swagger:response string
 type APIString string
 
+// APIRepoArchivedError is error that is raised when a repo is archived
```
Comment (delvh) [inline modules/context/api.go]:
```suggestion
// APIRepoArchivedError is an error that is raised when an archived repo should be modified
```

File: modules/context/api.go [hunk 5]
```diff
@@ -101,6 +101,12 @@ type APIRedirect struct{}
...
+// swagger:response repoArchived
```
Comment (delvh) [inline modules/context/api.go]:
```suggestion
// swagger:response repoArchivedError
```

File: routers/api/v1/repo/branch.go [hunk 6]
```diff
@@ -117,17 +117,13 @@ func DeleteBranch(ctx *context.APIContext) {
 	//     "$ref": "#/responses/error"
 	//   "404":
 	//     "$ref": "#/responses/notFound"
-
+	//   "423":
+	//     "$ref": "#/responses/repoArchived"
```
Comment (delvh) [inline routers/api/v1/repo/branch.go]: Please also add a description, here and everywhere else:
```suggestion
	//     "$ref": "#/responses/repoArchived"
	//    "description": "The repo is archived"
```

File: routers/api/v1/api.go [hunk 7]
```diff
@@ -675,7 +675,7 @@ func mustEnableWiki(ctx *context.APIContext) {
 
 func mustNotBeArchived(ctx *context.APIContext) {
 	if ctx.Repo.Repository.IsArchived {
-		ctx.NotFound()
+		ctx.Error(http.StatusLocked, "RepoArchived", fmt.Errorf("repo %s/%s is archived", ctx.Repo.Repository.OwnerName, ctx.Repo.Repository.Name))
```
Comment (delvh) [inline routers/api/v1/api.go]: I think you can use `LogString` here as well.
However, I don't know if on `Repo` or on `Repository`.

File: models/repo/repo.go (hunk 2 above)
Comment (delvh) [reply models/repo/repo.go]: Preferably `MustNotBeArchived`

File: routers/api/v1/repo/branch.go (hunk 6 above)
Comment (JakobDev) [reply routers/api/v1/repo/branch.go]: The description is added by swagger automatically

**Review by delvh — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Pull request merged**